services/device-registry-base/src/main/java/org/eclipse/hono/service/management/credentials/CommonCredential.java

```diff
@@ -164,4 +167,41 @@ public void checkValidity() {
     public CommonCredential stripPrivateInfo() {
         return this;
     }
+
+    /**
+     * Merges the secrets of the given credential with that of the current one based on the secret ids.
+     *
+     * @param credential The credential to be merged.
+     * @return a reference to this for fluent use.
+     * @throws IllegalArgumentException if the given credential is invalid and cannot be merged.
+     * @throws NullPointerException if the given credential is {@code null}.
+     */
+    @JsonIgnore
+    public CommonCredential merge(final CommonCredential credential) {
+
+        Objects.requireNonNull(credential);
+
+        if (!getType().equals(credential.getType())) {
+            throw new IllegalArgumentException("credential to be merged must be of the same type");
+        }
+
+        getSecrets()
+                .forEach(secret -> Optional.ofNullable(secret.getId())
+                        .ifPresent(secretId -> findSecretById(credential, secretId)
+                                .ifPresentOrElse(secret::merge, () -> {
+                                    throw new IllegalArgumentException(
+                                            String.format("secret [id: %s] not found", secret.getId()));
+                                })));
+
+        return this;
+    }
+
+
+    private Optional<? extends CommonSecret> findSecretById(final CommonCredential credential, final String secretId) {
```

Comment:
FMPOV we do not need the *credentials* parameter, do we?

Reply:
This `credentials` parameter is needed. It is the credential that is received as a parameter for merging in `CommonCredential merge(final CommonCredential credential)`

Reply:
sure, but the other object also is a `CommonCredential`, isn't it? So the code in the *merge* operation would simply need to be changed to:
```java
getSecrets()
                .forEach(secret -> Optional.ofNullable(secret.getId())
                        .ifPresent(secretId -> credential.findSecretById(secretId)
                                .ifPresentOrElse(secret::merge, () -> {
                                    throw new IllegalArgumentException(
                                            String.format("secret [id: %s] not found", secret.getId()));
                                })));
```
or am I mistaken?

Reply:
Now I understood what you meant :-)

Reply:
I have made those changes now.